ros/src/waypoint_updater/jmt.py

```diff
@@ -0,0 +1,170 @@
+#!/usr/bin/env python
+
+import rospy
+import math
+import numpy as np
+
+
+class JMT(object):
+    def __init__(self, start, end, T):
+
+        """
+        Calculates Jerk Minimizing Trajectory for start, end and T.
+        start and end include
+        [displacement, velocity, acceleration]
+        """
+        self.start = start
+        self.end = end
+        self.final_displacement = end[0]
+        self.T = T
+
+        a_0, a_1, a_2 = start[0], start[1], start[2] / 2.0
+        c_0 = a_0 + a_1 * T + a_2 * T**2
+        c_1 = a_1 + 2 * a_2 * T
+        c_2 = 2 * a_2
+
+        A = np.array([
+                     [T**3,   T**4,    T**5],
+                     [3*T**2, 4*T**3,  5*T**4],
+                     [6*T,   12*T**2, 20*T**3],
+                     ])
+
+        B = np.array([
+                     end[0] - c_0,
+                     end[1] - c_1,
+                     end[2] - c_2
+                     ])
+        a_3_4_5 = np.linalg.solve(A, B)
+        self.coeffs = np.concatenate([np.array([a_0, a_1, a_2]), a_3_4_5])
+
+    # def JMTD_at(self, displacement, coeffs, t0, tmax, deq_wpt_ptr):
+    def JMTD_at(self, displacement, t0, tmax):
+        # find JMT descriptors at displacement
+        s_last = 0.0
+        t_found = False
+        t_inc = 0.01
+        iterations = 0
+
+        for t_cnt in range(int((tmax-t0)/t_inc) + 100):
+            iterations += 1
+            t = t0 + t_cnt * t_inc
+            s = self.get_s_at(t)
+            if s > displacement:
+                t = t - (1 - (s - displacement) / (s - s_last)) * t_inc
+                t_found = True
+                break
+            # end if
+            s_last = s
+        # end for
+        if t_found is False:
+            rospy.loginfo("waypoint_updater:JMTD_at Ran out of bounds without "
+                          "finding target displacement")
+            return None
+
+        s = self.get_s_at(t)
+        delta_s = (displacement - s)
+        if delta_s > 0.0:
+            searchdir = 1.0
+        else:
+            searchdir = -1.0
+
+        t_smallinc = searchdir * 0.005
+        while delta_s * searchdir > 0.0:
+            iterations += 1
+            t += t_smallinc
+            delta_s = displacement - self.get_s_at(t)
```

Comment:
This is actually a pretty efficient, but I would like to try the spline method, also because the s value might be more accurate than the value calculated using linear distance between nodes.